qp_klp/klp.py

```diff
@@ -32,32 +44,228 @@ def sequence_processing_pipeline(qclient, job_id, parameters, out_dir):
     run_identifier = parameters.pop('run_identifier')
     sample_sheet = parameters.pop('sample_sheet')
+    job_pool_size = 30
+
+    # checking if this is running as part of the unittest
+    # https://stackoverflow.com/a/25025987
+    skip_exec = True if [x for x in stack() if
+                         'unittest' in x.filename] else False
 
     success = True
     ainfo = None
     msg = None
-    if exists(run_identifier) and isdir(run_identifier):
-        if {'body', 'content_type', 'filename'} == set(sample_sheet):
-            outpath = partial(join, out_dir)
 
-            with open(outpath('listing.txt'), 'w') as f:
-                f.write('\n'.join(glob(join(run_identifier, '*'))))
+    qclient.update_job_step(job_id, "Step 1 of 6: Setting up pipeline")
 
-            with open(outpath(sample_sheet['filename']), 'w') as f:
-                f.write(sample_sheet['body'])
+    if {'body', 'content_type', 'filename'} == set(sample_sheet):
+        outpath = partial(join, out_dir)
+        final_results_path = outpath('final_results')
+        makedirs(final_results_path, exist_ok=True)
 
-            ainfo = [
-                ArtifactInfo('output', 'job-output-folder',
-                             [(f'{out_dir}/', 'directory')])
-            ]
+        # the user is uploading a sample-sheet to us, but we need the
+        # sample-sheet as a file to pass to the Pipeline().
+        sample_sheet_path = outpath(sample_sheet['filename'])
+        with open(sample_sheet_path, 'w') as f:
+            f.write(sample_sheet['body'])
+
+        # validate the sample-sheet using metapool package.
+        sheet = KLSampleSheet(sample_sheet_path)
+        val_sheet = validate_and_scrub_sample_sheet(sheet)
+        if not val_sheet:
+            qclient.update_job_step(job_id,
+                                    "Sample sheet failed validation.")
+            raise ValueError("Sample sheet failed validiation")
         else:
-            success = False
-            msg = "Doesn't look like a valid uploaded file; please review."
+            # get project names and their associated qiita ids
+            bioinformatics = val_sheet.Bioinformatics
+            lst = bioinformatics.to_dict('records')
+
+        # find the uploads directory all trimmed files will need to be
+        # moved to.
+        results = qclient.get("/qiita_db/artifacts/types/")
+
+        # trimmed files are stored by qiita_id. Find the qiita_id
+        # associated with each project and ensure a subdirectory exists
+        # for when it comes time to move the trimmed files.
+        special_map = []
+        for result in lst:
+            project_name = result['Sample_Project']
+            qiita_id = result['QiitaID']
+            upload_path = join(results['uploads'], qiita_id)
+            makedirs(upload_path, exist_ok=True)
+            special_map.append((project_name, upload_path))
+
+        config_fp = ('/home/qiita_test/qiita-spots/'
```

Comment:
No more hardcoded paths please!!

Reply:
Sorry, my bad. It's an old line that snuck back in during merging. It's been removed. Pipeline() should use the config_fp referred to in the actions file.